desc/singularities.py

```diff
@@ -0,0 +1,730 @@
+"""High order method for singular surface integrals, from Malhotra 2019."""
+from abc import ABC, abstractmethod
+
+import numpy as np
+import scipy
+from interpax import fft_interp2d
+
+from desc.backend import fori_loop, jnp, put, vmap
+from desc.basis import DoubleFourierSeries
+from desc.compute import rpz2xyz, rpz2xyz_vec
+from desc.io import IOAble
+from desc.utils import isalmostequal, islinspaced
+
+
+def safediv(a, b, fill=0, threshold=0):
+    """Divide a/b with guards for division by zero.
+
+    Parameters
+    ----------
+    a, b : ndarray
+        Numerator and denominator.
+    fill : float, ndarray, optional
+        Value to return where b is zero.
+    threshold : float >= 0
+        How small is b allowed to be.
+    """
+    mask = jnp.abs(b) <= threshold
+    num = jnp.where(mask, fill, a)
+    den = jnp.where(mask, 1, b)
+    return num / den
+
+
+def safenorm(x, ord=None, axis=None, fill=0, threshold=0):
+    """Like jnp.linalg.norm, but without nan gradient at x=0.
+
+    Parameters
+    ----------
+    x : ndarray
+        Vector or array to norm.
+    ord : {non-zero int, inf, -inf, ‘fro’, ‘nuc’}, optional
+        Order of norm.
+    axis : {None, int, 2-tuple of ints}, optional
+        Axis to take norm along.
+    fill : float, ndarray, optional
+        Value to return where x is zero.
+    threshold : float >= 0
+        How small is x allowed to be.
+
+    """
+    is_zero = (jnp.abs(x) <= threshold).all(axis=axis, keepdims=True)
+    x = jnp.where(is_zero, jnp.ones_like(x), x)  # replace x with ones if is_zero
+    n = jnp.linalg.norm(x, ord=ord, axis=axis)
+    n = jnp.where(is_zero.squeeze(), fill, n)  # replace norm with zero if is_zero
+    return n
+
+
+def _get_quadrature_nodes(q):
+    """Polar nodes for quadrature around singular point.
+
+    Parameters
+    ----------
+    q : int
+        Order of quadrature in radial and azimuthal directions.
+
+    Returns
+    -------
+    r, w : ndarray
+        Radial and azimuthal coordinates
+    dr, dw : ndarray
+        Radial and azimuthal spacing/quadrature weights
+
+    """
+    Nr = Nw = q
+    r, dr = scipy.special.roots_legendre(Nr)
+    # integrate separately over [-1,0] and [0,1]
+    r1 = 1 / 2 * r - 1 / 2
+    r2 = 1 / 2 * r + 1 / 2
+    r = jnp.concatenate([r1, r2])
+    dr = jnp.concatenate([dr, dr]) / 2
+    w = np.linspace(0, np.pi, Nw, endpoint=False)
+    dw = jnp.ones_like(w) * np.pi / Nw
+    r, w = jnp.meshgrid(r, w)
+    r = r.flatten()
+    w = w.flatten()
+    dr, dw = jnp.meshgrid(dr, dw)
+    dr = dr.flatten()
+    dw = dw.flatten()
+    return r, w, dr, dw
+
+
+class _BIESTInterpolator(IOAble, ABC):
```

Comment:
Do you have your own writeup of this stuff (like for your free bdry paper)? Would be nice to have (for me) when reviewing this file